src/authentication/authentication-hapi-plugin.ts

```diff
@@ -0,0 +1,154 @@
+
+  constructor(
+    @inject(AuthenticationModule.injectSymbol) authenticationModule: AuthenticationModule,
+    @inject(authServerBaseUrlInjectSymbol) authServerBaseUrl: string,
+    @inject(loggerInjectSymbol) logger: Logger,
+    @inject(GitlabClient.injectSymbol) gitlab: GitlabClient,
+    @inject(jwtOptionsInjectSymbol) jwtOptions: auth.JWTStrategyOptions,
+    ) {
+    super({
+      name: 'authentication-plugin',
+      version: '1.0.0',
+    });
+    this.authenticationModule = authenticationModule;
+    this.authServerBaseUrl = authServerBaseUrl;
+    this.logger = logger;
+    this.gitlab = gitlab;
+    this.options = jwtOptions;
+  }
+
+  public async register(server: Hapi.Server, _options: Hapi.IServerOptions, next: () => void) {
+    await this.registerAuth(server);
+    server.route([{
+      method: 'GET',
+      path: '/signup/{teamToken}',
+      handler: {
+        async: this.signupHandler,
+      },
+      config: {
+        bind: this,
+        validate: {
+          params: {
+            teamToken: Joi.string().alphanum().length(8),
+          },
+        },
+      },
+    },
+    ]);
+    server.route([{
+      method: 'GET',
+      path: '/team',
+      handler: {
+        async: this.getTeamHandler,
+      },
+      config: {
+        bind: this,
+      },
+    }]);
+    next();
+  }
+
+  public async signupHandler(request: Hapi.Request, reply: Hapi.IReply) {
+    const credentials = request.auth.credentials as AccessToken;
+    const teamToken = credentials[this.teamTokenClaimKey];
+
+    reply(teamToken);
+  }
+
+  public async getTeamHandler(request: Hapi.Request, reply: Hapi.IReply) {
+    const credentials = request.auth.credentials as AccessToken;
+    const user = await getUserByEmail(credentials[this.subEmailClaimKey], this.gitlab);
+    const teams = getUserTeams(user.id, this.gitlab);
+    return reply(teams);
+  }
+
+  public async signToTeam(_uid: UserId, teamToken: string) {
+    return teamToken;
+  }
+
+  public async getTeam(uid: UserId) {
+    return uid.idp;
+  }
+
+  public async validateUser(
+    payload: AccessToken,
+    _request: any,
+    callback: (err: any, valid: boolean, credentials?: any) => void,
+  ) {
+    // NOTE: we have just a single team at this point so no further checks are necessary
+    callback(undefined, true, payload);
+  }
+
+  public async registerAuth(server: Hapi.Server) {
+    await server.register(auth);
+    server.auth.strategy('jwt', 'jwt', true, {
+      ...this.options,
+      validateFunc: this.validateUser.bind(this),
+    });
+  }
+}
+
+export default AuthenticationHapiPlugin;
+
+export function parseSub(sub: string) {
+  if (typeof sub !== 'string') {
+    throw new Error('Invalid \'sub\' claim.');
+  }
+  const parts = sub.split('|');
+  if (parts.length !== 2) {
+    throw new Error('Invalid \'sub\' claim.');
+  }
+  return {
+    id: parts[1],
+    idp: parts[0],
+  };
+}
+
+export async function getUserByEmail(email: string, gitlab: GitlabClient) {
```

Comment:
Me too, but for circular dependency reasons it's not doable. I'd leave it like this for now and refactor more thoroughly in a separate PR.